### PR TITLE
Add initial version of ergo-installer script

### PR DESCRIPTION
**Pull request by andyceo**

### CODE REVIEW

**Comment by andyceo** [comment]
This PR address issues #970 #971.

Design notes:

1. No need root priveleges. Script print installation tips instead, that user supposed to execute manually
2. Script uses user's home dir and install Ergo to ~/Ergo by default.
3. Script always run Ergo node in daemon mode with `daemon` utility.
4. Script designed to be verbose and try to output current steps and settings.
5. For now script tested only on Ubuntu and works well.

@todo 

1. Tor support
2. Maybe some polishing

I propose writing any suggestions in this PR.

**Comment by andyceo** [comment]
Tipical output:

```
$ ./ergo-installer.sh --daemon --tor --api-key=donotlook
Ergo node with config file will be installed into '/home/andyceo/Ergo' directory and will be named as 'andyceo-karella' and has API key 'donotlook'.
Ergo directory '/home/andyceo/Ergo' already exists.
Latest known Ergo release: v3.1.3, downloading it to /home/andyceo/Ergo/ergo.jar with overwrite...
Ergo was downloaded to /home/andyceo/Ergo/ergo.jar
Executing ergo node to obtain API key hash.......
API key hash obtained: 0e5751c026e543b2e8ab2eb06099daa1d1e5df47778f7787faab45cdf12fe3a8
Stopping Ergo node with PID=6841....
Stopped.
Config file writed to /home/andyceo/Ergo/application.conf
Starting node.......
Node started.
To stop the node later, use: kill 6919
```

**Comment by andyceo** [comment]
So some extra design notes:

6. Script install node for Mainnet only
7. Tor support with binding API and Network to 127.0.0.1 if Tor used
8. It is also possible to run script from Github: `curl -s https://raw.githubusercontent.com/ergoplatform/ergo/937022adb19b2c09ce4074eb6914e6e88fb5b348/ergo-installer.sh | sh -s -- --api-key=donotlook`. We should change URL when this PR would be merged.

File: ergo-installer.sh [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+# Initialize all the option variables.
+# This ensures we are not contaminated by variables from the environment.
+APP_DIR=~/Ergo
+NODE_NAME=${USER}-$(hostname)
+API_KEY=
+INSTALLATION_RECOMMENDATIONS=
+TOR=no
+CONFIG_TEMPLATE="ergo.node.mining = false\nscorex.network.nodeName = \"<NODE_NAME>\"\nscorex.restApi.apiKeyHash = \"<API_KEY_HASH>\""
+NODE_PARAMS="-Xmx$(awk '/MemTotal/ { printf "%.0f", $2*0.9 }' /proc/meminfo)K"
```
Comment (aslesarenko) [inline ergo-installer.sh]: @andyceo please add comment to explaining the formula for heap size

File: ergo-installer.sh [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+if [ "${MODE}" != 'full' ] && [ "${MODE}" != 'mining' ]; then
+  echo 'ERROR: "--mode" requires a non-empty option argument that can be "full" or "mining" only.'
+  exit 1
+fi
+
+echo "Ergo node with config file will be installed into '${APP_DIR}' directory and will be named as '${NODE_NAME}' and has API key '${API_KEY}'."
+
+
+## Check preinstalled software
+INSTALLATION_RECOMMENDATIONS=
```
Comment (aslesarenko) [inline ergo-installer.sh]: @andyceo add additional check of operating system and fail fast when it is not supported by the script.

File: ergo-installer.sh [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+ERGO_DOWNLOAD_URL=https://github.com/ergoplatform/ergo/releases/download/${LATEST_ERGO_RELEASE}/ergo-${LATEST_ERGO_RELEASE_NUMBERS}.jar
+echo "Latest known Ergo release: ${LATEST_ERGO_RELEASE}, downloading it to ${APP_DIR}/ergo.jar with overwrite..."
+curl --silent -L ${ERGO_DOWNLOAD_URL} --output ${APP_DIR}/ergo.jar
+echo "Ergo was downloaded to ${APP_DIR}/ergo.jar"
+
+# First run of Ergo node, to create API key hash that later be added into config
+echo -n "Executing ergo node to obtain API key hash..."
+daemon --chdir=${APP_DIR} -- java -jar ${APP_DIR}/ergo.jar --mainnet
+PID=$(pgrep -f "daemon --chdir=${APP_DIR} -- java")
+while ! curl --output /dev/null --silent --head --fail http://localhost:9053; do sleep 1 && echo -n '.'; done;
```
Comment (aslesarenko) [inline ergo-installer.sh]: @andyceo please comment on what happens in this loop and the other loops

File: ergo-installer.sh (hunk 1 above)
Comment (andyceo) [reply ergo-installer.sh]: Done

File: ergo-installer.sh (hunk 3 above)
Comment (andyceo) [reply ergo-installer.sh]: Done

File: ergo-installer.sh (hunk 2 above)
Comment (andyceo) [reply ergo-installer.sh]: It is not possible as I can say for now... Script designed to be runnable on wide range of OSes. So if OS has shell (even Windows 10 has, but I not tested script on W10) and basic popular GNU software installed, script will download and run Ergo node.

Installation recommendations done for Debian-based systems, as I only sure how to install related software on Ubuntu/Debian, and that was done only for example. It is possible to add more recommendations for MacOS or Windows or other popular OSes, PRs are welcome!

So I coclude - if script run, than it executes as expected on almost every OS.

**Pull request merged**